tests/transaction_test_util.go

```diff
@@ -79,62 +80,78 @@ func (tt *TransactionTest) Run(config *params.ChainConfig) error {
-		{"Byzantium", types.NewEIP155Signer(config.ChainID), tt.Result["Byzantium"], true, false, false},
-		{"Constantinople", types.NewEIP155Signer(config.ChainID), tt.Result["Constantinople"], true, false, false},
-		{"Istanbul", types.NewEIP155Signer(config.ChainID), tt.Result["Istanbul"], true, true, false},
-		{"Berlin", types.NewEIP2930Signer(config.ChainID), tt.Result["Berlin"], true, true, false},
-		{"London", types.NewLondonSigner(config.ChainID), tt.Result["London"], true, true, false},
-		{"Paris", types.NewLondonSigner(config.ChainID), tt.Result["Paris"], true, true, false},
-		{"Shanghai", types.NewLondonSigner(config.ChainID), tt.Result["Shanghai"], true, true, true},
-		{"Cancun", types.NewCancunSigner(config.ChainID), tt.Result["Cancun"], true, true, true},
-		{"Prague", types.NewPragueSigner(config.ChainID), tt.Result["Prague"], true, true, true},
+		{"Frontier", false},
+		{"Homestead", false},
+		{"EIP150", false},
+		{"EIP158", false},
+		{"Byzantium", false},
+		{"Constantinople", false},
+		{"Istanbul", false},
+		{"Berlin", false},
+		{"London", false},
+		{"Paris", true},
+		{"Shanghai", true},
+		{"Cancun", true},
+		{"Prague", true},
 	} {
-		if testcase.fork == nil {
+		expected := tt.Result[testcase.name]
+		if expected == nil {
 			continue
 		}
-		sender, hash, gas, err := validateTx(tt.Txbytes, testcase.signer, testcase.isHomestead, testcase.isIstanbul, testcase.isShanghai)
+		config, ok := Forks[testcase.name]
+		if !ok || config == nil {
+			return UnsupportedForkError{Name: testcase.name}
+		}
+		var (
+			rules  = config.Rules(new(big.Int), testcase.isMerge, 0)
+			signer = types.MakeSigner(config, new(big.Int), 0)
+		)
+		sender, hash, gas, err := validateTx(tt.Txbytes, signer, &rules)
 		if err != nil {
-			if testcase.fork.Hash != nil {
-				return fmt.Errorf("unexpected error: %v", err)
+			if expected.Hash != nil {
+				return fmt.Errorf("unexpected error fork %s: %v", testcase.name, err)
 			}
 			continue
 		}
-		if testcase.fork.Exception != nil {
-			return fmt.Errorf("expected error %v, got none (%v)", *testcase.fork.Exception, err)
+		if expected.Exception != nil {
+			return fmt.Errorf("expected error %v, got none (%v), fork %s", *expected.Exception, err, testcase.name)
 		}
-		if common.Hash(*testcase.fork.Hash) != hash {
-			return fmt.Errorf("hash mismatch: got %x, want %x", hash, common.Hash(*testcase.fork.Hash))
+		if common.Hash(*expected.Hash) != hash {
+			return fmt.Errorf("hash mismatch: got %x, want %x", hash, common.Hash(*expected.Hash))
 		}
-		if common.Address(*testcase.fork.Sender) != sender {
-			return fmt.Errorf("sender mismatch: got %x, want %x", sender, testcase.fork.Sender)
+		if common.Address(*expected.Sender) != sender {
+			return fmt.Errorf("sender mismatch: got %x, want %x", sender, expected.Sender)
 		}
-		if hash != common.Hash(*testcase.fork.Hash) {
-			return fmt.Errorf("hash mismatch: got %x, want %x", hash, testcase.fork.Hash)
+		if hash != common.Hash(*expected.Hash) {
```

Comment:
Isn't this the same condition as four lines up? hash != *expected.Hash

Reply:
😄